dwave/plugins/sklearn/transformers.py

```diff
@@ -0,0 +1,292 @@
+# Copyright 2023 D-Wave Systems Inc.
+#
+#    Licensed under the Apache License, Version 2.0 (the "License");
+#    you may not use this file except in compliance with the License.
+#    You may obtain a copy of the License at
+#
+#        http://www.apache.org/licenses/LICENSE-2.0
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS,
+#    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#    See the License for the specific language governing permissions and
+#    limitations under the License.
+
+from __future__ import annotations
+
+import itertools
+import logging
+import tempfile
+import typing
+import warnings
+
+import dimod
+import numpy as np
+import numpy.typing as npt
+
+from dwave.cloud.exceptions import ConfigFileError, SolverAuthenticationError
+from dwave.system import LeapHybridCQMSampler
+from sklearn.base import BaseEstimator
+from sklearn.feature_selection import SelectorMixin
+from sklearn.utils.validation import check_is_fitted
+
+from dwave.plugins.sklearn.utilities import corrcoef
+
+__all__ = ["SelectFromQuadraticModel"]
+
+
+class SelectFromQuadraticModel(SelectorMixin, BaseEstimator):
+    """Select features using a quadratic optimization problem solved on a hybrid solver.
+
+    Args:
+        alpha:
+            Hyperparameter between 0 and 1 that controls the relative weight of
+            the relevance and redundancy terms.  `alpha=1` places all weight on
+            relevance and selects all features, whereas `alpha=0` places all
+            weight on redundancy and selects no features.
```

Comment:
This description is unclear to me; maybe it needs the context of sci-kit? I naively understand maximum "redundancy" (alpha=0) to mean that we accept any amount of redundancy in the information content of our input features and so make use of all input features (select all versus "and selects no features"). Similarly, for alpha=1 (preference for relevance) I'd expect the selection of a small subset of features that has no overlap in information content among the subset members and between any member and the non-selected features.

Reply:
This text is adopted from [our example](https://github.com/dwave-examples/feature-selection-cqm/blob/30fb675e85c4061f40334de4f2e96081f47061c3/correlation.py#L42-L46). I think it's fairly clear but if we want to change it, probably we should change it there as well.

Reply:
This is super confusing to me. When it says it says "selects all features" that seems like it should have an effect on the final number of features, but it doesn't obviously. When is says it selects no features same thing. I think the important distinction is that alpha as a hyper-parameter has a different meaning in BQM vs CQM, and those docs are for BQM. In the unconstrained context, a 0 alpha will give you 0 linear biases and then all positive quadratic biases, so the minimum energy is the all 0 solution, whereas the non-zero alpha will give you negative linear biases, so the model will select more features if the cross correlations aren't massive. In the CQM doesn't have that same dynamic because we fix the number of features selected.

Reply:
> it should have an effect on the final number of features

I was just trying to figure that same thing out now myself, so agree this is confusing:

```
In [21]: selector.fit_transform(in_123, out.reshape(5,), num_features=2, alpha=0)
Out[21]:
array([[ 0.10743986,  0.17592027],
       [-0.7199656 ,  0.3435838 ],
       [ 0.15304165,  0.7558532 ],
       [ 1.22055433,  1.52715311],
       [-0.19839271,  2.19253674]])

In [22]: selector.fit_transform(in_123, out.reshape(5,), num_features=2, alpha=1)
Out[22]:
array([[ 0.10743986,  0.17592027],
       [-0.7199656 ,  0.3435838 ],
       [ 0.15304165,  0.7558532 ],
       [ 1.22055433,  1.52715311],
       [-0.19839271,  2.19253674]])
```

Reply:
Ok, I will take a crack at rewriting.

Reply:
Ok, rewritten.
```
        alpha:
            Hyperparameter between 0 and 1 that controls the relative weight of
            the relevance and redundancy terms.
            ``alpha=0`` places no weight on the quality of the features,
            therefore the features will be selected as to minimize the
            redundancy without any consideration to quality.
            ``alpha=1`` places the maximum weight on the quality of the features,
            and therefore will be equivalent to using
            :class:`sklearn.feature_selection.SelectKBest`.
```

Reply:
I also updated the actual behavior of `alpha` and added tests, so it should behave as described.

Reply:
Much clearer!
One minor thing you might consider is the use of both "relevance" and "quality", rather than just one term (the cited paper uses "influence").  I prefer "relevance" (closer to predictiveness) rather than "quality", which is vague. But this seems clear to me.